### PR TITLE
Logical link tests updated

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
Tests added where nodes are connected to nodes through logical links and both consumers and producers are connected with logical links to nodes.

### CODE REVIEW

File: tests/models/unit_cases/case_3a/model/3a_cons2node.esdl [hunk 1]
```diff
@@ -0,0 +1,184 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (KobusVanRooyen) [inline tests/models/unit_cases/case_3a/model/3a_cons2node.esdl:1]: When I load the esdl in the mapeditor nothing is displayed. Usually this means that it is not a 100% valid esdl file. Can you please check this?

File: tests/models/unit_cases/case_3a/model/3a_prod2node2cons.esdl [hunk 2]
```diff
@@ -0,0 +1,146 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (KobusVanRooyen) [inline tests/models/unit_cases/case_3a/model/3a_prod2node2cons.esdl:1]: When I load the esdl in the mapeditor nothing is displayed. Usually this means that it is not a 100% valid esdl file. Can you please check this?

File: tests/models/unit_cases/case_3a/model/3a_prod2node.esdl [hunk 3]
```diff
@@ -0,0 +1,260 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Comment (KobusVanRooyen) [inline tests/models/unit_cases/case_3a/model/3a_prod2node.esdl:1]: When I load the esdl in the mapeditor nothing is displayed. Usually this means that it is not a 100% valid esdl file. Can you please check this?

File: tests/test_logical_links.py [hunk 4]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+        """
+        This test checks if the logic implemented for a partially logically linked energy system.
+        Meaning an energy system where assets are directly connected to each other without a network
+        (transport asset(s)) in between. This specific test covers the logical connection between
+        producers and consumers to nodes in a heat network. Please note that due to the closed
+        system constraints we need to set the minimum velocity to zero to avoid backflow in the
+        solution.
+
+        This is the most basic check where we have a simple network and check for the basic physics.
+        This simple network includes one source, some pipes and logical links, node, a tank storage
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py:182]: When I read this text it is not clear that there are 2 different setups/test. I suggest to indicate this more clearly. Something like:

Setup 1: only producer and demand

Setup 2: more advance with storage.

File: tests/test_logical_links.py [hunk 5]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+        - results of 3a do not vary a lot.
+        """
+
+        import models.source_pipe_split_sink.src.double_pipe_heat as example
+        from models.source_pipe_split_sink.src.double_pipe_heat import SourcePipeSink
+
+        base_folder = Path(example.__file__).resolve().parent.parent
+
+        problem = run_esdl_mesido_optimization(
+            SourcePipeSink,
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py:206]: for this problem I would add a test to ensure : "minimum velocity has been to zero". So that if someone changes the setting for some reason, then it is directly checked. Not everyone will read the docstring in detail if the test fails due to this.

File: tests/test_logical_links.py [hunk 6]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+        production_list = []
+        for esdl_file in [
+            "3a_cons2node.esdl",
+            "3a_prod2node.esdl",
+            "3a_prod2node2cons.esdl",
+            "3a.esdl",
+        ]:
+            # Just a "problem is not infeasible"
+            heat_problem = run_esdl_mesido_optimization(
+                HeatProblem,
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py:243]: for this problem I would add a test to ensure : "minimum velocity has been to zero". So that if someone changes the setting for some reason, then it is directly checked. Not everyone will read the docstring in detail if the test fails due to this.

File: tests/test_logical_links.py [hunk 7]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
                 )
             np.testing.assert_allclose(discharge_sum, 0.0, atol=1.0e-12)
 
+    def test_logical_links_all_connections_heat(self):
+        """
+        This test checks if the logic implemented for a partially logically linked energy system.
+        Meaning an energy system where assets are directly connected to each other without a network
+        (transport asset(s)) in between. This specific test covers the logical connection between
+        producers and consumers to nodes in a heat network. Please note that due to the closed
+        system constraints we need to set the minimum velocity to zero to avoid backflow in the
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py]: I assume this (set the minimum velocity to zero) means setting self.heat_network_settings["minimum_velocity"]=0.0, but I do not see it in the problem setups. Or I might be misunderstanding the text?

File: tests/test_logical_links.py [hunk 8]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+        This simple network includes one source, some pipes and logical links, node, a tank storage
+        and 3 demands.
+        The buffer cannot yet be connected with a logical link, due to the flow direction variable.
+        The logical links can give scaling issues when used at demands.
+
+        Checks;
+        - Demand matching
+        - Energy conservation
+        - Heat to discharge
+        Check that the history for the buffer is set correctly at t=0
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py]: bullet indicator missing

File: tests/test_logical_links.py [hunk 9]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+            profile_reader=ProfileReaderFromFile,
+            input_timeseries_file="timeseries_import.csv",
+        )
+        results = problem.extract_results()
+
+        demand_matching_test(problem, results)
+        energy_conservation_test(problem, results)
+        heat_to_discharge_test(problem, results)
+
+        # Test conservatioon of flow at the nodes
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py]: What is the purpose of the comment? I find it a bit confusing

File: tests/test_logical_links.py [hunk 10]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
                 )
             np.testing.assert_allclose(discharge_sum, 0.0, atol=1.0e-12)
 
+    def test_logical_links_all_connections_heat(self):
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py:174]: can you add this a.test_logical_links_all_connections_heat() to main at the bottom of this script. I still use this setup unfortunately ;-)

File: tests/test_logical_links.py [hunk 11]
```diff
@@ -171,6 +171,117 @@ def test_logical_links_nodes_heat(self):
...
+                    np.sum(results[f"{buffer}.Heat_buffer"][1:] * 3600.0)
+                    - np.sum(results[f"{buffer}.Heat_loss"][1:] * 3600.0),
+                    atol=1.0,
+                )
+                np.testing.assert_allclose(results[f"{buffer}.Heat_buffer"][0], 0.0, atol=1.0e-6)
+
+        # check that the total heat production is almost the same for all models, with and without
+        # logical links.
+        total_productions = [sum(production_list[i]) for i in range(len(production_list))]
+        np.testing.assert_allclose(total_productions[-1], total_productions, rtol=1e-5)
```
Comment (KobusVanRooyen) [inline tests/test_logical_links.py]: you are comparing total_productions[-1] with itself as well. You could have had:

np.testing.assert_allclose(total_productions[-1], total_productions[:-1], rtol=1e-5)

But it is a small case, so not a big deal ;-)

File: tests/test_warmingup_unit_cases.py [hunk 12]
```diff
@@ -170,10 +170,6 @@ def test_3a(self):
             np.testing.assert_allclose(
                 results[f"{buffer}.Heat_buffer"], results[f"{buffer}.Heat_flow"]
             )
-            np.testing.assert_allclose(
```
Comment (KobusVanRooyen) [inline tests/test_warmingup_unit_cases.py:173]: I am not sure why this has been removed if it was working?

File: tests/models/unit_cases/case_3a/model/3a_cons2node.esdl (hunk 1 above)
Comment (FJanssen-TNO) [reply tests/models/unit_cases/case_3a/model/3a_cons2node.esdl:1]: Interesting, I actually drew those in the mapeditor...

File: tests/test_warmingup_unit_cases.py (hunk 12 above)
Comment (FJanssen-TNO) [reply tests/test_warmingup_unit_cases.py:173]: This is already part of the default tests in heat_to_discharge test or energy conservation.

File: tests/models/unit_cases/case_3a/model/3a_cons2node.esdl (hunk 1 above)
Comment (FJanssen-TNO) [reply tests/models/unit_cases/case_3a/model/3a_cons2node.esdl:1]: "flowRate" attribute was assigned to GeothermalSource in 3a.esdl and variants of it. This has now been removed.

File: tests/models/unit_cases/case_3a/model/3a_prod2node2cons.esdl (hunk 2 above)
Comment (FJanssen-TNO) [reply tests/models/unit_cases/case_3a/model/3a_prod2node2cons.esdl:1]: fixed, see above

File: tests/models/unit_cases/case_3a/model/3a_prod2node.esdl (hunk 3 above)
Comment (FJanssen-TNO) [reply tests/models/unit_cases/case_3a/model/3a_prod2node.esdl:1]: fixed, see above

File: tests/test_logical_links.py (hunk 5 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py:206]: This part of the docstring was copied and should not have been there. This is updated.

File: tests/test_logical_links.py (hunk 6 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py:243]: This part of the docstring was copied and should not have been there. This is updated.

File: tests/test_logical_links.py (hunk 7 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py]: This part of the docstring was copied and should not have been there. This is updated.

File: tests/test_logical_links.py (hunk 9 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py]: It explains the check below. Mass balance at the nodes, as now there are only logical links this extra check is usefull.

File: tests/test_logical_links.py (hunk 8 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py]: Updated

File: tests/test_logical_links.py (hunk 10 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py:174]: Updated it for you.

File: tests/test_logical_links.py (hunk 11 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py]: Yes indeed, I don't  mind, but I updated it for you.

File: tests/test_logical_links.py (hunk 4 above)
Comment (FJanssen-TNO) [reply tests/test_logical_links.py:182]: Clarified in a different way.

**Comment by FJanssen-TNO** [comment]
@KobusVanRooyen ready for review again.

**Review by KobusVanRooyen — APPROVED** [review]

**Comment by KobusVanRooyen** [comment]
@FJanssen-TNO approved
